src/Storages/PostgreSQL/PostgreSQLReplicationHandler.cpp

```diff
@@ -0,0 +1,458 @@
+{
+    if (Poco::File(metadata_path).exists())
+        Poco::File(metadata_path).remove();
+
+    connection = std::make_shared<PostgreSQLConnection>(connection_str);
+    auto tx = std::make_shared<pqxx::nontransaction>(*connection->conn());
+
+    dropPublication(tx);
+    if (isReplicationSlotExist(tx, replication_slot))
+        dropReplicationSlot(tx);
+
+    tx->commit();
+}
+
+
+NameSet PostgreSQLReplicationHandler::fetchRequiredTables(PostgreSQLConnection::ConnectionPtr connection_)
+{
+    if (tables_list.empty())
+    {
+        return fetchPostgreSQLTablesList(connection_);
+    }
+    else
+    {
+        createPublicationIfNeeded(connection_);
+        return fetchTablesFromPublication(connection_);
+    }
+}
+
+
+NameSet PostgreSQLReplicationHandler::fetchTablesFromPublication(PostgreSQLConnection::ConnectionPtr connection_)
+{
+    std::string query = fmt::format("SELECT tablename FROM pg_publication_tables WHERE pubname = '{}'", publication_name);
+    std::unordered_set<std::string> tables;
+    pqxx::read_transaction tx(*connection_);
+
+    for (auto table_name : tx.stream<std::string>(query))
+        tables.insert(std::get<0>(table_name));
+
+    return tables;
+}
+
+
+PostgreSQLTableStructure PostgreSQLReplicationHandler::fetchTableStructure(
+        std::shared_ptr<pqxx::work> tx, const std::string & table_name)
+{
+    auto use_nulls = context->getSettingsRef().external_table_functions_use_nulls;
+    return fetchPostgreSQLTableStructure(tx, table_name, use_nulls, true);
+}
+
+
+std::unordered_map<Int32, String> PostgreSQLReplicationHandler::reloadFromSnapshot(
+        const std::vector<std::pair<Int32, String>> & relation_data)
+{
+    std::unordered_map<Int32, String> tables_start_lsn;
+    try
+    {
+        auto tx = std::make_shared<pqxx::work>(*connection->conn());
+        Storages sync_storages;
+        for (const auto & relation : relation_data)
+        {
+            const auto & table_name = relation.second;
+            auto * storage = storages[table_name];
+            sync_storages[table_name] = storage;
+            storage->dropNested();
+        }
+        tx->commit();
+
+        auto replication_connection = std::make_shared<PostgreSQLConnection>(fmt::format("{} replication=database", connection_str));
+        replication_connection->conn()->set_variable("default_transaction_isolation", "'repeatable read'");
+
+        auto ntx = std::make_shared<pqxx::nontransaction>(*replication_connection->conn());
+        std::string snapshot_name, start_lsn;
+        createReplicationSlot(ntx, start_lsn, snapshot_name, true);
+        /// This snapshot is valid up to the end of the transaction, which exported it.
+        auto success_tables = loadFromSnapshot(snapshot_name, sync_storages);
+        ntx->commit();
+
+        for (const auto & relation : relation_data)
+        {
+            if (success_tables.find(relation.second) != success_tables.end())
```

Comment:
Seems to be always true (see the comment about return value of `loadFromSnapshot(...)`)

Reply:
Fixed.